painttiming.bs

```diff
@@ -62,9 +87,34 @@ Formally, we consider the user agent to have "rendered" a document when it has p
 
     NOTE: The rendering pipeline is very complex, and the timestamp should be the latest timestamp the user agent is able to note in this pipeline (best effort). Typically the time at which the frame is submitted to the OS for display is recommended for this API.
 
+<dfn>Element or pseudo-element</dfn>: A [=document-sourced element=] or a [=generated content pseudo-element=].
+<dfn>Contentful</dfn>: An [=element or pseudo-element=] counts as contentful when one or more of the following apply:
+* The element has a [=text node=] child, with at least one character, excluding [=document white space characters=].
+* The element is a [=replaced image=] representing a [=loaded=] image.
+* The element's [=computed=] style contains a [=loaded=] [=background-image=].
+* The element is a [=canvas=] with its [=context mode=] set to any value other than <code>none</code>.
+* The element is an [=svg element with rendered descendants=].
+* The element's [=computed=] style contains a [=contentful=] and [=paintable=] [=generated content pseudo-element=].
+
+<dfn>Paintable</dfn>: As a general rule, an [=element or pseudo-element=] is paintable if it is within the viewport, or can potentially be in the viewport as a result of scrolling.
+An [=element or pseudo-element=] counts as paintable when all of the following apply:
+* The element's [=computed=] [=visibility=] is <code>visible</code>.
+* The element and all of its ancestors have positive [=computed=] [=opacity=].
+
+    NOTE: there could be cases where a <code>paintable</code> [=element or pseudo-element=] would not be visible to the user, for example in the case of text that has the same color as its background.
+    Those elements would still considered as paintable for the purpose of computing [=first contentful paint=].    
+
+* The element's [=paintable bounding rect=] has positive right, bottom, width and height values. 
+    
+    NOTE: This covers the cases where the element is scaled to zero size, has <code>display: none</code>, or <code>display: contents</code> where the contents resolve to an empty rect.
+
+<dfn>Paintable Bounding Rect</dfn>: The paintable bounding rect is the element's bounding rect, which would be the result of running the [=getBoundingClientRect=] algorithm on the [=element or pseudo-element=], clipped by ancestors with <code>overflow: clip</code>.
```

Comment:
[`getBoundingClientRect`](https://drafts.csswg.org/cssom-view/#dom-element-getboundingclientrect) isn't defined on a [pseudo element](https://drafts.csswg.org/css-pseudo-4/#CSSPseudoElement-interface)?

Reply:
right. I don't know of a good model to find whether a pseudo-element rect is paintable. Since it's an edge-case of an edge-case, I'll try to find something simple that maybe doesn't cover all the cases.

Reply:
Hm... indeed. We may need to refactor [CSS OM view module](https://drafts.csswg.org/cssom-view/) so that these bounding box concepts can be defined for a pseudo element.